lib/oauth2/client.rb

```diff
@@ -258,7 +258,17 @@ def authenticator
     #
     # @return [AccessToken] the initialized AccessToken
     def build_access_token(response, access_token_opts, access_token_class)
-      access_token_class.from_hash(self, response.parsed.merge(access_token_opts)).tap do |access_token|
+      token_opts = response.parsed.merge(access_token_opts)
+
+      access_token =
+        if access_token_class == OAuth2::MACToken
+          token = AccessToken.from_hash(self, token_opts)
+          OAuth2::MACToken.from_access_token(token, secret)
```

Comment:
I think `secret` needs to be something else according to https://tools.ietf.org/id/draft-ietf-oauth-v2-http-mac-02.html.